### PR TITLE
feat: session handling refactoring. Step 1

### PR DESCRIPTION
**Pull request by sadym-chromium**
1. Add `browser.close` command support.
2. Don't close browser when WS connection is closed.
3. Close browser after each e2e test.

Out of scope:
1. Reconnecting to existing BiDi sessions.

### CODE REVIEW

File: src/bidiServer/index.ts [hunk 1]
```diff
@@ -34,6 +34,8 @@ import {BidiServerRunner, debugInfo} from './bidiServerRunner.js';
 import {MapperServer} from './MapperServer.js';
 import {readMapperTabFile} from './reader.js';
 
+export type CloseBrowserDelegate = () => Promise<void>;
```
Comment (thiagowfx) [inline src/bidiServer/index.ts:37]: Document this

File: tests/conftest.py [hunk 2]
```diff
@@ -30,6 +30,10 @@ async def websocket():
     url = f"ws://localhost:{port}"
     async with websockets.connect(url) as connection:
         yield connection
+        await execute_command(connection, {
+            "method": "browser.close",
+            "params": {}
+        })
```
Comment (thiagowfx) [inline tests/conftest.py:36]: This does not need `try` / `except`? (just asking)

File: src/bidiServer/bidiServerRunner.ts [hunk 3]
```diff
@@ -149,8 +150,37 @@ export class BidiServerRunner {
...
+
+        // Try to parse the message to handle some of BiDi commands.
+        let parsedCommandData: {id: number; method: string};
+        try {
+          parsedCommandData = JSON.parse(plainCommandData);
+        } catch (e) {
+          this.#respondWithError(
+            connection,
+            {},
+            'invalid argument',
```
Comment (thiagowfx) [inline src/bidiServer/bidiServerRunner.ts:163]: Reuse `ErrorCode.InvalidArgument` from src/protocol/webdriver-bidi.ts

File: src/bidiServer/bidiServerRunner.ts [hunk 4]
```diff
@@ -149,8 +150,37 @@ export class BidiServerRunner {
...
+            `Cannot parse data as JSON`
```
Comment (thiagowfx) [inline src/bidiServer/bidiServerRunner.ts:164]: Print the data as well?

Comment (sadym-chromium) [reply src/bidiServer/bidiServerRunner.ts:164]: It is printed [here](https://github.com/GoogleChromeLabs/chromium-bidi/blob/20b8b8fb8ac08298c5cb1d81ff7e6e4cb7ede7a4/src/bidiServer/bidiServerRunner.ts#L153).

File: src/bidiServer/bidiServerRunner.ts (hunk 3 above)
Comment (sadym-chromium) [reply src/bidiServer/bidiServerRunner.ts:163]: done

File: src/bidiServer/index.ts (hunk 1 above)
Comment (sadym-chromium) [reply src/bidiServer/index.ts:37]: done

**Review by thiagowfx — APPROVED** [review]

File: src/bidiServer/bidiServerRunner.ts [hunk 5]
```diff
@@ -129,39 +134,67 @@ export class BidiServerRunner {
...
+      connection.on('close', async () => {
         debugInternal(
           `${new Date().toString()} Peer ${
             connection.remoteAddress
           } disconnected.`
         );
-
-        onBidiConnectionClosed();
+        // TODO: handle reconnection which is used in WPT. Until then, close the
+        //  browser after eachWS connection is closed.
```
Comment (sadym-chromium) [inline src/bidiServer/bidiServerRunner.ts:196]:
```suggestion
        //  browser after each WS connection is closed.
```

**Comment by OrKoN** [comment]
did you mean to mark a refactoring as `feat:`?

**Comment by thiagowfx** [comment]
Indeed, it should have been `chrore:` or similar. I tend not to pay much attention to those, but will keep it in mind in future reviews.

**Comment by OrKoN** [comment]
better to use `refactor:` for refactorings

**Pull request merged**